// @slyte/core/src/service.js
import { __getCurrentContext, __getCurrentFunc, __setCurrentContext, toLowerCase, Lyte } from "@slyte/core/src/lyte.js"
import { extendEventListeners } from "@slyte/core/src/lyte-utils";
/*convert to custom class*/
class Service { 
    constructor() {
        var context = __getCurrentContext();
        // if(context){
        if(context.__isAddon) {
            if(this.__isAddon){
                this.$parent = context;
                this.$app = context.$app;
            }else{
                this.$addon = context;
            }
            if(toLowerCase(this.constructor.name) !== "utils"){
                this.$utils = context.$utils;
            }
        } else {
            this.$app = context;
            if(this.__isAddon){
                this.$parent = context;
            }
            if(toLowerCase(this.constructor.name) !== "utils"){
                this.$utils = context.$utils;
            }
        }
        var func = __getCurrentFunc();
        var lookupMap;
        if(this.__isAddon){
            Object.defineProperty(this, "__lyte", {
                enumerable : false, 
                writable : true, 
                value : {"lookupMap" : new Map()}
            });
            lookupMap = this.__lyte.lookupMap;
            extendEventListeners(this.__lyte.lookupMap);
        }
        else{
            if(this.$addon){
                lookupMap = this.$addon.__lyte.lookupMap;
            }else if(this.$app){
                lookupMap = this.$app.__lyte.lookupMap;
            }
        }
        if(func) {
            func(this);
        }
        var oldContext = __getCurrentContext();
        if(this.__isAddon) {
            __setCurrentContext(this);
        }
        
        if(lookupMap.get(this.constructor)){
            lookupMap.set(this.constructor,this);
        }

        var lookups = this.lookups();
        // if(this.$addon){ //commented since lookups returned array - we are not using or holding it for anny purpose
        //     this.$addon.__lyte.lookupMap.lookups = lookups;
        // }else{
        //     this.$app.__lyte.lookupMap.lookups = lookups;
        // }
        __handleLookups(lookups, this.__isAddon ? this : context, this);
        this.__lookups = lookups;
        if(this.__isAddon) {
            __setCurrentContext(oldContext);
        }
        // }else{
            // debugger
        // }
    }
    
    lookups() {
        return [];
    }
}
function __getLyte(service){
    return service.$addon ? service.$addon.__lyte : service.__lyte ? service.__lyte : service.$app.__lyte;
}
function __handleLookups(lookups, app, curr) {
    if(lookups && lookups.length){
        var appLookupMap = app.__lyte.lookupMap;
        var currLookupMap = __getLyte(curr).lookupMap;
        lookups.forEach(function(item) {
            if(!item){
                Lyte.error("Invalid class/instance passed in "+ curr.constructor.name + "'s lookup hook.")
                return
            }
            if(typeof item == "string") {
                if(curr) {
                    if(item.length){
                        if(appLookupMap.get(item)){
                            curr["$" + item] = appLookupMap.get(item);
                        }
                        else{
                            var id = app.__lyte.lookupMap.addEventListener(item, function(ins){
                                curr["$" + item] = ins;
                                app.__lyte.lookupMap.removeEventListener(id);
                            });
                        }
                    }
                    else{
                        Lyte.error("Lookup cannot be added for empty(undefined) key");
                    }
                }
            } else {
                if(__checkIfService(item)) {
                    instantiateService(item, app, curr);
                } else if(Object.keys(item).length == 1) {
                    for(var onlyKey in item){
                        var val = item[onlyKey];
                        if(onlyKey.length){
                            if(/^(utils|node|lg|app|addon)$/.test(onlyKey)){
                                Lyte.error("Lookup cannot be added for the predefined key - "+ onlyKey);
                                return;
                            }
                            let name = "$" + onlyKey;
                            if(__checkIfService(val)) {
                                instantiateService(val, app, curr, name);
                            } else {
                                if(val && typeof val == "object"){
                                    curr[name] = val;
                                    if(!appLookupMap.get(onlyKey)){
                                        appLookupMap.set(onlyKey, val);
                                        appLookupMap.triggerEvent(onlyKey, val);
                                    }
                                }
                                else if(typeof val == "string"){
                                    if(appLookupMap.get(val)){
                                        curr[name] = appLookupMap.get(val);
                                    }
                                    else{
                                        var id = appLookupMap.addEventListener(val, function(ins){
                                            curr[name] = ins;
                                            appLookupMap.removeEventListener(id);
                                        });
                                    }
                                }
                                currLookupMap.set(onlyKey, val);
                            }
                        }
                        else{
                            Lyte.error("Lookup cannot be added for empty(undefined) key");
                        }
                    }
                }
            }
            
        });
    }
}

function instantiateService(cls, app, curr, name) {
    name = name || ("$" + toLowerCase(cls.name));
    let appLookupMap = app.__lyte.lookupMap;
    // let currLookupMap = curr.__lyte.lookupMap;
    let lookupValue = appLookupMap.get(cls);
    if(lookupValue) {
        if(lookupValue instanceof Promise) {
            lookupValue.then(function(val) {
                curr[name] = val;
            });
        } else {
            curr[name] = lookupValue;
        }
    } else {
        appLookupMap.set(cls, cls);
        let obj = curr[name] = new cls();
        appLookupMap.triggerEvent(name, obj);
    }
}
function __checkIfService(item) {
    var isService = false;
    while(item != null) {
        if(item == Service || item.__lyteOrigClass) {
            isService = true;
            break;
        } else {
            item = Object.getPrototypeOf(item);
        }
    }
    return isService;
}
export { Service, __checkIfService, __handleLookups ,__getLyte}
